module/os/linux/spl/spl-taskq.c

```diff
@@ -1306,16 +1302,14 @@ spl_taskq_expand(unsigned int cpu, struct hlist_node *node)
 	if (!(tq->tq_flags & TASKQ_ACTIVE))
 		goto out;
 
-	if (tq->tq_flags & TASKQ_THREADS_CPU_PCT) {
-		int nthreads = MIN(tq->tq_cpu_pct, 100);
-		nthreads = MAX(((num_online_cpus() + 1) * nthreads) / 100, 1);
-		tq->tq_maxthreads = nthreads;
-	} else {
-		tq->tq_maxthreads++;
-	}
+	ASSERT(tq->tq_flags & TASKQ_THREADS_CPU_PCT);
+	int nthreads = MIN(tq->tq_cpu_pct, 100);
+	nthreads = MAX(((num_online_cpus() + 1) * nthreads) / 100, 1);
+	tq->tq_maxthreads = nthreads;
 
 	if (!((tq->tq_flags & TASKQ_DYNAMIC) && spl_taskq_thread_dynamic) &&
 	    tq->tq_maxthreads > tq->tq_nthreads) {
+		ASSERT3U(tq->tq_maxthreads, ==, tq->tq_nthreads + 1);
```

Comment:
Since `spl_taskq_thread_dynamic` can be changed at run time I don't think it's safe to ASSERT this here.

Reply:
It looks like you resolved this by removing the assertion.  So the logic is that if the desired number of threads differs from the actual number of threads by more than one, we'll only create/destroy one thread.  Wouldn't we want to make the number of threads equal to the desired number?  e.g. by turning the `if` into a `while` loop?

Reply:
I guess the question I'm asking is "which code is responsible for growing/shrinking the taskq if spl_taskq_thread_dynamic is turned off at runtime?" That code should already be doing this. If there is no such code, then I guess we can have it happen here, but it's only going to run when CPUs are added or removed, which feels incorrect to me. Is the fix to have this code take on that responsibility, or to have other code that runs more regularly do so, or to simply continue to not have such logic?

Reply:
Right now there isn't any code that modifies the taskq's threads if `spl_taskq_thread_dynamic ` is changed.  It's simply a writable module option which can be changed at any time, i.e `echo 0 >/sys/module/spl/parameters/spl_taskq_thread_dynamic`.  What I was trying to avoid is someone toggling the value, say while investigating some performance issue, and then unexpectedly hitting this ASSERT latter.  We could add code which is run when the module option is changed, but that feels unnecessary to me since the option was just added to help us assess the performance of static vs dynamic taskqs.  This isn't something which should ever really happen, I just want to make sure that if it does it isn't fatal and we end up in a sane state.